gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java

```diff
@@ -0,0 +1,239 @@
+/**
+ * A Google cloud storage object.
+ */
+public final class Blob {
+
+  private final Storage storage;
+  private BlobInfo info;
+
+  public static class BlobSourceOption extends Option {
+
+    private static final long serialVersionUID = 214616862061934846L;
+
+    private BlobSourceOption(StorageRpc.Option rpcOption) {
+      super(rpcOption, null);
+    }
+
+    private Storage.BlobSourceOption convert(BlobInfo blobInfo) {
+      switch (rpcOption()) {
+        case IF_GENERATION_MATCH:
+          return Storage.BlobSourceOption.generationMatch(blobInfo.generation());
+        case IF_GENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.generationNotMatch(blobInfo.generation());
+        case IF_METAGENERATION_MATCH:
+          return Storage.BlobSourceOption.metagenerationMatch(blobInfo.metageneration());
+        case IF_METAGENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.metagenerationNotMatch(blobInfo.metageneration());
+        default:
+          throw new AssertionError("Unexpected enum value");
+      }
+    }
+
+    public static BlobSourceOption generationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_MATCH);
+    }
+
+    public static BlobSourceOption generationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_NOT_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_NOT_MATCH);
+    }
+
+    static Storage.BlobSourceOption[] convert(BlobInfo blobInfo, BlobSourceOption... options) {
+      Storage.BlobSourceOption[] convertedOptions = new Storage.BlobSourceOption[options.length];
+      int index = 0;
+      for (BlobSourceOption option : options) {
+        convertedOptions[index++] = option.convert(blobInfo);
+      }
+      return convertedOptions;
+    }
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided {@code BlobInfo}. The storage service is used
+   * to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param info blob's info
+   */
+  public Blob(Storage storage, BlobInfo info) {
+    this.storage = checkNotNull(storage);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided bucket and blob names. The storage service is
+   * used to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param bucket bucket's name
+   * @param blob blob's name
+   */
+  public Blob(Storage storage, String bucket, String blob) {
+    this.storage = checkNotNull(storage);
+    this.info = BlobInfo.of(checkNotNull(bucket), checkNotNull(blob));
+  }
+
+  /**
+   * Return the blob's information.
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
```

Comment:
I vote we go with returning stale info and having the reload function.  I think pinning a Blob to it's info would be fine, and I'd vote for forcing the version match by default to discourage situations like the `oldBlob`, `newBlob`, `newestBlob` Marco brought up.  I think pinning a Blob to it's info would probably be a little bit clearer for users because they're less likely to view it as a mutable, up-to-date copy of what's stored in GCS and write buggy code due to version issues.

I don't know if this is relevant/correct, but by pinning a version, it seems like the Blob class would just provide syntactic sugar.  Was that the original point, or was there additional functionality that we lose by pinning a Blob to it's BlobInfo?